### ISSUE TITLE
HTTPDownloader doesn't fully inherit from HTTPClientFactory

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2971]
|<img alt="paulegan's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| paulegan reported|
|-|-|
|Trac ID|trac#2971|
|Type|defect|
|Created|2008-01-03 18:50:11Z|
|Branch|https://github.com/twisted/twisted/tree/httpdownloader-fixes-2971|

The `gotHeaders` method of the `HTTPDownloader` class should set `self.response_headers` and `self.cookies`, just as `HTTPClientFactory` does.

Attachments:

* [twisted_web_client-download_headers.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/76e/76e14d2b128dec91b6e21196aa9a680392eded87/f22c0b349d9d86b8fa4fa6887adb6cc316ad8415.diff) (450 bytes) - added by paulegan on 2008-01-03 18:50:45Z - Patch to call parent method to update headers
* [web_client-downloader.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/76e/76e14d2b128dec91b6e21196aa9a680392eded87/3eb9cfaf6d77d57840036c39749b2517757d1c47.patch) (3943 bytes) - added by paulegan on 2008-01-20 16:24:20Z - Added some basic tests

<details><summary>Searchable metadata</summary>

```
trac-id__2971 2971
type__defect defect
reporter__paulegan paulegan
priority__normal normal
milestone__ 
branch__branches_httpdownloader_fixes_2971 branches/httpdownloader-fixes-2971
branch_author__paulegan__exarkun paulegan, exarkun
status__closed closed
resolution__fixed fixed
component__web web
keywords__httpclient httpclient
time__1199386211000000 1199386211000000
changetime__1232111233000000 1232111233000000
version__None None
owner__ 
cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="paulegan's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|paulegan commented|
|-|-|

As well as setting the response headers, `HTTPDownloader` should support the `timeout`, `cookies`, and `followRedirect` arguments.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve set owner to paulegan|
|-|-|

Are you willing to write unittests for these changes?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

(In [25772]) Branching to 'httpdownloader-fixes-2971'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

(In [25773]) Apply web_client-downloader.patch, minus parts which don't apply but are apparently unrelated anyway

refs [#2971](https://github.com/twisted/twisted/issues/2971)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

  1. Coding standard issues:
    1. Tests should be named like `test_foo` rather than `testFoo`.  `testDownloadTimeout` should be `test_downloadTimeout` and `testDownloadHeaders` should be `test_downloadHeaders`.
    1. Test methods should have docstrings explaining the behavior being tested.
  1. Testing issues
    1. There's no test coverage for the added support for cookies
    1. `twisted.web.test.test_webclient.WebClientSSLTestCase.testDownloadHeaders` fails with a `ConnectionDone` error
    1. Rather than changing the behavior of the "wait" resource in order to test some new behavior, a new resource with the new behavior should be added.  This way all the existing tests continue to test what they were testing and the new behavior can be tested as well.
  1. Functional issues
    1. `HTTPClientFactory` has a new `redirectLimit` parameter which `HTTPDownloader` still doesn't support.  This suggests that there should be a more general way to inherit the functionality of `HTTPClientFactory`, perhaps using `**kwargs` in `__init__`.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

(In [25774]) use _makeGetterFactory in testDownloadHeaders is valid for HTTPS as well


refs [#2971](https://github.com/twisted/twisted/issues/2971)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

(In [25775]) test behavior of timeouts for the case of some data received and the case of no data received

refs [#2971](https://github.com/twisted/twisted/issues/2971)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [25776]) add a test for the cookies parameter to `client.HTTPDownloader.__init__`

refs [#2971](https://github.com/twisted/twisted/issues/2971)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

(In [25777]) Fix new test method names and give them docstrings

refs [#2971](https://github.com/twisted/twisted/issues/2971)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

(In [25778]) add redirectLimit support to HTTPDownloader

refs [#2971](https://github.com/twisted/twisted/issues/2971)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

(In [25779]) update copyright dates

refs [#2971](https://github.com/twisted/twisted/issues/2971)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

All review points addressed, I think.  Unfortunately I didn't find a more general way to inherit functionality from `HTTPClientFactory` in `HTTPDownloader`.  There is an extra parameter to `HTTPDownloader.__init__` in between parameters supported by `HTTPClientFactory.__init__` which messes things up.  So I just added `redirectLimit` as another special case.  Hopefully the new HTTP client will obsolete all of this code soon anyway.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_14"></a>|@glyph set owner to @exarkun|
|-|-|

`ForeverTakingResource` could use a docstring.  But other than that, great!  Go ahead and merge it.  (I hope you're right about the new client.)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun commented|
|-|-|

(In [26038]) docstring for `ForeverTakingResource`

refs [#2971](https://github.com/twisted/twisted/issues/2971)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun set status to `closed`|
|-|-|

(In [26040]) Merge httpdownloader-fixes-2971

Author: paulegan, exarkun
Reviewer: glyph
Fixes: [#2971](https://github.com/twisted/twisted/issues/2971)

Implement cookie, redirect limits, and timeouts for `downloadPage`.  These are the
same as the features already available for `getPage`.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
